app/scripts/modules/core/src/artifact/expectedArtifact.service.ts

```diff
@@ -0,0 +1,45 @@
+import { copy, module } from 'angular';
+
+import { IArtifact, IExpectedArtifact, IPipeline, IStage } from 'core/domain';
+import { PipelineConfigService } from 'core';
+
+export class ExpectedArtifactService {
+  constructor(private pipelineConfigService: PipelineConfigService) {
+    'ngInject';
+  }
+
+  public summarizeExpectedArtifact(expected: IExpectedArtifact): string {
+    if (!expected) {
+      return '';
+    }
+
+    const artifact = copy(expected.matchArtifact);
+    return Object.keys(artifact)
+      .filter((k: keyof IArtifact) => artifact[k])
+      .filter((k: keyof IArtifact) => k !== 'kind')
+      .map((k: keyof IArtifact) => (`${k}: ${artifact[k]}`))
+      .join(', ');
+  };
+
+  public getExpectedArtifactsAvailableToStage(stage: IStage, pipeline: IPipeline): IExpectedArtifact[] {
+    let result = pipeline.expectedArtifacts || [];
+    this.pipelineConfigService.getAllUpstreamDependencies(pipeline, stage)
+      .forEach(s => {
+        const expectedArtifact = (s as any)['expectedArtifact'];
```

Comment:
`const expectedArtifact: IExpectedArtifact = (s as any).expectedArtifact;`, can drop the cast below.